File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <stack>
#include <algorithm>
using namespace std;

struct Time{
    string s;
    int t;
    int re;
};

bool cmp(Time &t1, Time &t2){
    return t1.t < t2.t;
}

int tTOm(string s){
    string a = "";
    for(int i=0;i<2;i++)
        a += s[i];
    
    int sum = stoi(a) * 60;
    
    a = "";
    for(int i= 3;i<5;i++)
        a += s[i];
    
    sum += stoi(a);
    
    return sum;
}

vector<string> solution(vector<vector<string>> plans) {
    vector<string> answer;
    
    vector<Time> time(plans.size());
    stack<Time> st;
    
    for(int i=0;i<plans.size();i++){
        time[i].t = tTOm(plans[i][1]);
        time[i].s = plans[i][0];
        time[i].re = stoi(plans[i][2]);
    }
    
    sort(time.begin(),time.end(),cmp);
    
    for(int i=0;i<plans.size()-1;i++){
        int t = time[i].t + time[i].re;
        if(t > time[i+1].t){
            st.push({time[i].s, time[i+1].t, t-time[i+1].t});
        }
        else if(t == time[i+1].t)
            answer.push_back(time[i].s);
        else if(t < time[i+1].t){
            answer.push_back(time[i].s);
            
            while(st.size() !=0){
                Time T = st.top();
                T.t = t;
                st.pop();
                
                t = T.re + T.t;
                if(t < time[i+1].t){
                    answer.push_back(T.s);
                }
                else if(t == time[i+1].t){
                    answer.push_back(T.s);
                    break;
                }
                else if(t > time[i+1].t){
                    st.push({T.s, time[i+1].t, t-time[i+1].t});
                    break;
                }
                
            }
        }
    }
    answer.push_back(time[time.size()-1].s);
    
    while(st.size()!=0){
        answer.push_back(st.top().s);
        st.pop();
    }
    
    return answer;
}

File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.java
import java.io.*;
import java.util.*;
class Solution {
    public String[] solution(String[][] plans) {
        int n = plans.length;
        String[] answer = new String[n];
        int index = 0;
        
        xy[] v = new xy[n];
        Stack<xy> st = new Stack<>();
        
        for(int i = 0; i < n; i++){
            v[i] = new xy();
            v[i].s = plans[i][0];
            int h = Integer.parseInt(plans[i][1].substring(0,2));
            int m = Integer.parseInt(plans[i][1].substring(3,5));
            int f = Integer.parseInt(plans[i][2]);
            v[i].st = h * 60 + m;
            v[i].remain = f;
        }
        Arrays.sort(v, (a, b) -> a.st - b.st);
        
        for(xy x : v){
            if(st.isEmpty()) {
                st.add(new xy(x.s, x.st, x.remain));
                continue;
            }
        
            while(!st.isEmpty()){
                int finish = st.peek().st + st.peek().remain;
                if(finish > x.st){  // 진행 중이던 과제가 끝나기 전 새과제
                st.peek().remain -= x.st - st.peek().st;
                st.peek().st = x.st;
                break;
                }
                // 새과제 스타트 전까지 진행중이던 과제 업데이트
                answer[index++] = st.pop().s;
                if(!st.isEmpty())
                    st.peek().st = finish;
            }
            st.add(new xy(x.s, x.st, x.remain));
        }
        while(!st.isEmpty()) answer[index++] = st.pop().s;
        
        
        return answer;
    }
    static class xy{
        String s;
        int st, remain;
        xy(){}
        xy(String a, int b, int c){
            s = a; st = b; remain = c; 
        }
    }
}